src/main/java/com/github/adeynack/executors/LimitedExecutor.java

```diff
@@ -0,0 +1,92 @@
+package com.github.adeynack.executors;
+
+import org.jetbrains.annotations.NotNull;
+
+import javax.annotation.concurrent.GuardedBy;
+import java.util.ArrayDeque;
+import java.util.Queue;
+import java.util.concurrent.Executor;
+import java.util.concurrent.locks.ReentrantLock;
+
+@SuppressWarnings("WeakerAccess") // Library class
+public class LimitedExecutor implements Executor {
+
+  private final ReentrantLock mainLock = new ReentrantLock();
+
+  private final Executor baseExecutor;
+
+  private final int parallelTaskLimit;
+
+  @GuardedBy("mainLock")
+  private final Queue<Runnable> taskQueue = new ArrayDeque<>();
+
+  /**
+   * indicates the number of tasks already submitted to the underlying {@link Executor}.
+   */
+  @GuardedBy("mainLock")
+  private volatile int tasksSubmitted = 0;
+
+  public LimitedExecutor(Executor baseExecutor, int parallelTaskLimit) {
+    this.baseExecutor = baseExecutor;
+    this.parallelTaskLimit = parallelTaskLimit;
+  }
+
+  @Override
+  public void execute(@NotNull Runnable command) {
+    mainLock.lock();
+    try {
+      if (tasksSubmitted >= parallelTaskLimit) {
+        taskQueue.add(command);
+      } else {
+        final int actualTasksSubmitted = tasksSubmitted;
+        tasksSubmitted = actualTasksSubmitted + 1;
+        baseExecutor.execute(createTaskWrapper(command));
+      }
+    } finally {
+      mainLock.unlock();
+    }
+  }
+
+  private Runnable createTaskWrapper(Runnable command) {
+    return () -> {
+      try {
+        command.run();
+      } finally {
+        mainLock.lock();
+        try {
+          Runnable nextTask = taskQueue.poll();
+          if (nextTask == null) {
+            final int actualTasksSubmitted = tasksSubmitted;
+            tasksSubmitted = actualTasksSubmitted - 1;
+          } else {
+            baseExecutor.execute(createTaskWrapper(nextTask));
+          }
+        } finally {
+          mainLock.unlock();
+        }
+      }
+    };
+  }
+
+  /**
+   * @return the current count of tasks submitted to the underlying {@link Executor}.
+   */
+  public int getTasksSubmitted() {
+    int actualTasksSubmitted;
+    mainLock.lock();
+    try {
+      actualTasksSubmitted = tasksSubmitted;
+    } finally {
+      mainLock.unlock();
+    }
+    return actualTasksSubmitted;
+  }
+
+  /**
+   * @return a copy of the current queue of tasks.
+   */
+  public Queue<Runnable> getTaskQueue() {
+    return new ArrayDeque<>(this.taskQueue);
+  }
+
+}
```

Comment:
Overall this approach will ofc work (if implemented correctly). 
Minor suggestion would be to move logic for incrementing tasksSubmitted and submitting tasks to one method "tryExecute" or something and than change logic of `execute` to put task in a queue anyway and then call "tryExecute".  Same thing after each task is finished - just call "tryExecute"

Major suggestion would be to try to come up with non-blocking implementation (just using non-blocking queue) to reduce thread contention and maximize performance

Reply:
I sketched possible non-blocking solution in Kotlin (haven't test it, it's just a raw idea):
```koltin
import java.util.concurrent.ConcurrentLinkedDeque
import java.util.concurrent.Executor
import java.util.concurrent.atomic.AtomicInteger

class LimitedExecutor(private val baseExecutor: Executor, private val parallelTaskLimit: Int) : Executor {
    private val taskQueue = ConcurrentLinkedDeque<Runnable>()
    private val tasksRunning = AtomicInteger()

    override fun execute(command: Runnable) {
        taskQueue.add(command)
        trySubmit()
    }

    private fun trySubmit() {
        val running = tasksRunning.get()
        if (running < parallelTaskLimit) {
            val task = taskQueue.poll()
            val res = tasksRunning.compareAndSet(running, running + 1)
            if (res) {
                baseExecutor.execute(createTaskWrapper(task))
            }
        }
    }

    private fun createTaskWrapper(command: Runnable): Runnable = Runnable {
        try {
            command.run()
        } finally {
            tasksRunning.decrementAndGet()
            trySubmit()
        }
    }

    fun getTasksSubmitted(): Int = tasksRunning.get()
}

```